src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java

```diff
@@ -144,15 +188,28 @@ class ZipFileSystem extends FileSystem {
         this.useTempFile  = isTrue(env, "useTempFile");
         this.forceEnd64 = isTrue(env, "forceZIP64End");
         this.defaultCompressionMethod = getDefaultCompressionMethod(env);
+
+        AccessMode accessMode = AccessMode.from(env.get(PROPERTY_ACCESS_MODE));
+        boolean forceReadOnly = (accessMode == AccessMode.READ_ONLY);
+
         this.supportPosix = isTrue(env, PROPERTY_POSIX);
         this.defaultOwner = supportPosix ? initOwner(zfpath, env) : null;
         this.defaultGroup = supportPosix ? initGroup(zfpath, env) : null;
-        this.defaultPermissions = supportPosix ? initPermissions(env) : null;
+        this.defaultPermissions = supportPosix ? Collections.unmodifiableSet(initPermissions(env)) : null;
         this.supportedFileAttributeViews = supportPosix ?
-            Set.of("basic", "posix", "zip") : Set.of("basic", "zip");
+                Set.of("basic", "posix", "zip") : Set.of("basic", "zip");
+
+        // 'create=true' is semantically the same as StandardOpenOption.CREATE,
+        // and can only be used to create a writable file system (whether the
+        // underlying ZIP file exists or not), and is always incompatible with
+        // 'accessMode=readOnly').
+        boolean shouldCreate = isTrue(env, "create");
+        if (shouldCreate && forceReadOnly) {
+            throw new IllegalArgumentException(
```

Comment:
Although `IllegalArgumentException` seems reasonable here, the current contract of this constructor is to throw `IOException` and that then gets propagated through the public `FileSystemProvider.newFileSystem(...)` API which is specified to throw `IOException`.

So we will either have to throw `IOException` here (preferable) or we have to catch `IllegalArgumentException` at the call sites of this constructor and then rethrow it as a `IOException`, to prevent the unspecified `IllegalArgumentException` propagating out of the `FileSystemProvider.newFileSystem(...)` API.

Reply:
I checked the overloads of newFileSystem that were added in JDK 13 and somehow the `@throws IllegalArgumentException` was missed. It's declared by the methods that take a URI + env but not methods that take a Path + env. We should fix that, then the IAE from the zip provider won't be a surprise.

Reply:
I'll file an issue for this today, shortly.

Reply:
I created JDK-8356888 yesterday and [bplb](https://github.com/bplb) has picked it up already.